File: src/main/java/com/github/dwendelen/platformd/core/transaction/AccountBudget.java
package com.github.dwendelen.platformd.core.transaction;

import com.github.dwendelen.platformd.core.account.command.AcceptMoneyMade;
import com.github.dwendelen.platformd.core.account.command.RejectMoneyMade;
import com.github.dwendelen.platformd.core.account.event.MoneyMadeInitiated;
import com.github.dwendelen.platformd.core.budget.command.RecordIncome;
import com.github.dwendelen.platformd.core.budget.event.IncomeRecorded;
import org.axonframework.commandhandling.gateway.CommandGateway;
import org.axonframework.eventhandling.EventHandler;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import org.springframework.web.bind.annotation.ExceptionHandler;

@Component
public class AccountBudget {
    @Autowired
    private CommandGateway gateway;

    @EventHandler
    public void on(MoneyMadeInitiated event) {
        try {
            gateway.sendAndWait(new RecordIncome()
                    .setTransactionId(event.getTransactionId())
                    .setIncomeSource(event.getIncomeSource())
                    .setAmount(event.getAmount())
                    .setAccountId(event.getAccountId())
            );
        } catch (Exception e) {
            gateway.sendAndWait(new RejectMoneyMade()
                    .setAccountId(event.getAccountId())
                    .setTransactionId(event.getTransactionId())
                    .setReason(e.getMessage())
            );
        }
    }

    @EventHandler
    public void on(IncomeRecorded event) {
        gateway.sendAndWait(new AcceptMoneyMade()
                .setAccountId(event.getAccountId())
                .setTransactionId(event.getTransactionId())
        );
    }
}

File: application/src/main/scala/com/github/dwendelen/platformd/core/transaction/AccountBudget.scala
package com.github.dwendelen.platformd.core.transaction

import com.github.dwendelen.platformd.core.account.{AcceptMoneyMade, MoneyMadeInitiated, NormalAccount, RejectMoneyMade}
import com.github.dwendelen.platformd.core.budget.{IncomeRecorded, IncomeSource, RecordIncome}
import com.github.dwendelen.platformd.infrastructure.cqrs.{CQRS, Listener}
import org.springframework.stereotype.Component

@Component
class AccountBudget(cqrs: CQRS) extends Listener {
    cqrs.registerListener(this)

    def on(event: MoneyMadeInitiated): Unit = {
        try {
            cqrs.execute(
                event.incomeSource,
                new RecordIncome(
                    event.incomeSource,
                    event.transactionId,
                    event.amount,
                    event.accountId),
                classOf[IncomeSource]
            )
        } catch {
            case e: Exception =>
                cqrs.execute(event.accountId,
                    new RejectMoneyMade(
                        event.accountId,
                        event.transactionId,
                        e.getMessage),
                    classOf[NormalAccount]
                )
        }
    }

    def on(event: IncomeRecorded): Unit = {
        cqrs.execute(
            event.accountId,
            new AcceptMoneyMade(
                event.accountId,
                event.transactionId),
            classOf[NormalAccount]
        )
    }

    override def handle(event: Any): Unit = event match {
        case e: MoneyMadeInitiated => on(e)
        case e: IncomeRecorded => on(e)
        case _ =>
    }
}